docs/4_same_classes.md

```diff
@@ -0,0 +1,41 @@
+
+Using course description also introduces similar caveats as course titles do. In fact, professors are likely to make small modifications to their descriptions over time (for instance, changing the textbook used) that make exact matching suboptimal. Although we can use measures of fuzzy string similarity as a heuristic, we would ideally want to consider all three of code, title, and description when making our judgments.
+
+## Methodology
+
+Our input data source is a table of classes as well as one of listings – the `courses` and `listings` tables. The `courses` table stores a unique entry for each offering of a course, deduplicated by cross-listings (using cross-listing data provided by Yale), along with several attributes including title and description. The `listings` table stores the cross-listings of each course, which amounts to the code a course was listed in, as well as a pointer back to the deduplicated course in `courses`.
+
+The stated objective is to find all equivalent offerings for a given course. Since this amounts to evaluating an equivalence relationship on our set of courses, we can reformulate this task as computing a partition of our set of courses into groups of equivalent courses. Each group contains different offerings of the same course, and our set of groups is pairwise disjoint. Therefore, we just assign a unique group label to each course.
+
+Note that our previous strategy of identifying equivalent courses by shared code does not amount to an equivalence relationship – it's possible for courses *A*, B as well as *B*, *C* to have common codes but *A* and *C* to have no common ones, which violates the transitivity condition. This pattern sometimes occurs when course codes drift over time.
+
+To find these groups of equivalent courses, or equivalence classes, we model the set of courses as nodes in a graph, where each course is represented by a node. We then connect two nodes if and only if they are the same course. As a result, finding groups of equivalent courses amounts to finding [connected components](https://en.wikipedia.org/wiki/Component_(graph_theory)) in this graph, and it is easy to see that connected components are themselves a partition of the set of courses as desired. In fact, we previously used connected components when resolving cross-listings (sometimes Yale's cross-listings are incomplete).
+
+Therefore, our methodology is as follows:
+
+1. For each course in `courses`, create a node in the graph corresponding to the course.
+2. Connect the pairs of courses that share at least one code in common.
+3. Prune the edges to remove false positives. Given an edge between two courses *A* and *B*, keep it only if:
```

Comment:
You mention handling false positives. Are we doing anything to handle false negatives, or would that be too computationally intensive?

Reply:
Good point – the current method is primarily geared towards false positives, as we start by building the same-codes graph and pruning this. To address false negatives (i.e. failing to connect two courses that are the same) we would need to expand upon this shared-code matching. Doing pairwise comparisons of text distance between course titles or descriptions isn't feasible for initial search (as we're doing O(n^2) comparisons), but perhaps there's an application of the title/description embeddings to be realized here.